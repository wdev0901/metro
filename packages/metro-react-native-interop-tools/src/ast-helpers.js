/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+metro_bundler
 * @flow strict-local
 * @format
 */

import type {
  FunctionTypeAnnotation as BabelNodeFunctionTypeAnnotation,
  FunctionTypeParam as BabelNodeFunctionTypeParam,
  ObjectTypeAnnotation as BabelNodeObjectTypeAnnotation,
  ObjectTypeProperty as BabelNodeObjectTypeProperty,
  ObjectTypeSpreadProperty as BabelNodeObjectTypeSpreadProperty,
  TupleTypeAnnotation as BabelNodeTupleTypeAnnotation,
  NullableTypeAnnotation as BabelNodeNullableTypeAnnotation,
  GenericTypeAnnotation as BabelNodeGenericTypeAnnotation,
  UnionTypeAnnotation as BabelNodeUnionTypeAnnotation,
  IntersectionTypeAnnotation as BabelNodeIntersectionTypeAnnotation,
  ArrayTypeAnnotation as BabelNodeArrayTypeAnnotation,
  InterfaceExtends as BabelNodeInterfaceExtends,
  StringLiteralTypeAnnotation as BabelNodeStringLiteralTypeAnnotation,
  NumberLiteralTypeAnnotation as BabelNodeNumberLiteralTypeAnnotation,
} from '@babel/types';

import type {
  AnyTypeAnnotation,
  ArrayTypeAnnotation,
  NullableTypeAnnotation,
  FunctionTypeAnnotation,
  ObjectTypeAnnotation,
  FunctionTypeParam,
  ObjectTypeProperty,
  TupleTypeAnnotation,
  GenericTypeAnnotation,
  UnionTypeAnnotation,
  IntersectionTypeAnnotation,
  StringLiteralTypeAnnotation,
  NumberLiteralTypeAnnotation,
} from './type-annotation.js';

export type Schema = {
  typegenSchema: {},
  ...
};

export function isTurboModule(i: BabelNodeInterfaceExtends): boolean {
  return (
    i.id.name === 'TurboModule' &&
    (i.typeParameters == null || i.typeParameters.params.length === 0)
  );
}

export function getNodeLoc(
  loc: ?BabelNodeSourceLocation,
): ?BabelNodeSourceLocation {
  return loc == null
    ? null
    : {
        start: loc.start,
        end: loc.end,
      };
}

export function getTypeAnnotation(typeNode: BabelNodeFlow): AnyTypeAnnotation {
  switch (typeNode.type) {
    case 'BooleanTypeAnnotation':
    case 'NumberTypeAnnotation':
    case 'StringTypeAnnotation':
    case 'VoidTypeAnnotation':
    case 'AnyTypeAnnotation':
      return {
        type: typeNode.type,
        loc: getNodeLoc(typeNode.loc),
      };

    case 'NumberLiteralTypeAnnotation':
      return getNumberLiteralTypeAnnotation(typeNode);

    case 'StringLiteralTypeAnnotation':
      return getStringLiteralTypeAnnotation(typeNode);

    case 'ArrayTypeAnnotation':
      return getArrayTypeAnnotation(typeNode);

    case 'NullableTypeAnnotation':
      return getNullableTypeAnnotation(typeNode);

    case 'FunctionTypeAnnotation':
      return getFunctionTypeAnnotation(typeNode);

    case 'ObjectTypeAnnotation':
      return getObjectTypeAnnotation(typeNode);

    case 'TupleTypeAnnotation':
      return getTupleTypeAnnotation(typeNode);

    case 'GenericTypeAnnotation':
      return getGenericTypeAnnotation(typeNode);

    case 'UnionTypeAnnotation':
      return getUnionTypeAnnotation(typeNode);

    case 'IntersectionTypeAnnotation':
      return getIntersectionTypeAnnotation(typeNode);

    default:
      return {type: 'UnknownTypeAnnotation', loc: null};
  }
}

export function getFunctionTypeAnnotation(
  typeNode: BabelNodeFunctionTypeAnnotation,
): FunctionTypeAnnotation {
  return {
    type: 'FunctionTypeAnnotation',
    loc: getNodeLoc(typeNode.loc),
    params: typeNode.params.map(getFunctionTypeParameter),
    returnTypeAnnotation: getTypeAnnotation(typeNode.returnType),
  };
}

export function getFunctionTypeParameter(
  param: BabelNodeFunctionTypeParam,
): FunctionTypeParam {
  return {
    name: param.name?.name,
    typeAnnotation: getTypeAnnotation(param.typeAnnotation),
  };
}

export function getObjectTypeAnnotation(
  typeNode: BabelNodeObjectTypeAnnotation,
): ObjectTypeAnnotation {
  return {
    type: typeNode.type,
    loc: getNodeLoc(typeNode.loc),
    properties: typeNode.properties.map(
      (
        property:
          | BabelNodeObjectTypeProperty
          | BabelNodeObjectTypeSpreadProperty,
      ) => {
        return property.type === 'ObjectTypeProperty'
          ? getObjectTypeProperty(property)
          : getObjectTypeSpreadProperty(property);
      },
    ),
  };
}

//TODO T127639272 add support for spread properties
export function getObjectTypeSpreadProperty(
  typeProperty: BabelNodeObjectTypeSpreadProperty,
): ObjectTypeProperty {
  return {
    loc: getNodeLoc(typeProperty.loc),
    name: '',
    optional: false,
    typeAnnotation: {
      type: 'UnknownTypeAnnotation',
      loc: null,
    },
  };
}

export function getObjectTypeProperty(
  typeProperty: BabelNodeObjectTypeProperty,
): ObjectTypeProperty {
  return {
    loc: getNodeLoc(typeProperty.loc),
    name: getNameFromTypeProperty(typeProperty.key),
    optional: typeProperty.optional,
    typeAnnotation: getTypeAnnotation(typeProperty.value),
  };
}

export function getNameFromTypeProperty(
  node: BabelNodeIdentifier | BabelNodeStringLiteral,
): string {
  if (node.type === 'StringLiteral') {
    return node.value;
  }
  return node.name;
}

export function getTupleTypeAnnotation(
  typeNode: BabelNodeTupleTypeAnnotation,
): TupleTypeAnnotation {
  return {
    type: 'TupleTypeAnnotation',
    loc: getNodeLoc(typeNode.loc),
    types: typeNode.types.map(getTypeAnnotation),
  };
}

export function getNullableTypeAnnotation(
  typeNode: BabelNodeNullableTypeAnnotation,
): NullableTypeAnnotation {
  return {
    type: typeNode.type,
    loc: getNodeLoc(typeNode.loc),
    typeAnnotation: getTypeAnnotation(typeNode.typeAnnotation),
  };
}

export function getNameFromGenericNode(
  node: BabelNodeIdentifier | BabelNodeQualifiedTypeIdentifier,
): string {
  if (node.type === 'Identifier') {
    return node.name;
  }
  return node.id.name;
}

export function getGenericTypeAnnotation(
  typeNode: BabelNodeGenericTypeAnnotation,
): GenericTypeAnnotation {
  return {
    type: typeNode.type,
    loc: getNodeLoc(typeNode.loc),
    name: getNameFromGenericNode(typeNode.id),
    typeParameters: typeNode.typeParameters?.params
      ? typeNode.typeParameters.params?.map(getTypeAnnotation)
      : [],
  };
}

export function getUnionTypeAnnotation(
  typeNode: BabelNodeUnionTypeAnnotation,
): UnionTypeAnnotation {
  return {
    type: typeNode.type,
    loc: getNodeLoc(typeNode.loc),
    types: typeNode.types.map(getTypeAnnotation),
  };
}

export function getIntersectionTypeAnnotation(
  typeNode: BabelNodeIntersectionTypeAnnotation,
): IntersectionTypeAnnotation {
  return {
    type: typeNode.type,
    loc: getNodeLoc(typeNode.loc),
    types: typeNode.types.map(getTypeAnnotation),
  };
}

export function getArrayTypeAnnotation(
  typeNode: BabelNodeArrayTypeAnnotation,
): ArrayTypeAnnotation {
  return {
    type: typeNode.type,
    loc: getNodeLoc(typeNode.loc),
    elementType: getTypeAnnotation(typeNode.elementType),
  };
}

export function getStringLiteralTypeAnnotation(
  typeNode: BabelNodeStringLiteralTypeAnnotation,
): StringLiteralTypeAnnotation {
  return {
    type: typeNode.type,
    loc: getNodeLoc(typeNode.loc),
    value: typeNode.value,
  };
}

export function getNumberLiteralTypeAnnotation(
  typeNode: BabelNodeNumberLiteralTypeAnnotation,
): NumberLiteralTypeAnnotation {
  return {
    type: typeNode.type,
    loc: getNodeLoc(typeNode.loc),
    value: typeNode.value,
  };
}
