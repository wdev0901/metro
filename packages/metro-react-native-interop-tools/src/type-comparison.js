/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 * @oncall metro_bundler
 */

import type {
  AnyTypeAnnotation,
  NullableTypeAnnotation,
  LiteralTypeAnnotation,
} from './type-annotation.js';
import type {SourceLocation} from '@babel/types';

type ComparisonResult = $ReadOnly<{
  message: string,
  newTypeLoc: ?SourceLocation,
  oldTypeLoc: ?SourceLocation,
}>;

function removeNullable(annotation: NullableTypeAnnotation) {
  if (annotation.typeAnnotation.type === 'NullableTypeAnnotation') {
    return removeNullable(annotation.typeAnnotation);
  }
  return annotation.typeAnnotation;
}

function getSimplifiedType(annotation: AnyTypeAnnotation): string {
  switch (annotation.type) {
    case 'BooleanTypeAnnotation':
      return 'boolean';
    case 'StringTypeAnnotation':
      return 'string';
    case 'NumberTypeAnnotation':
      return 'number';
    case 'VoidTypeAnnotation':
      return 'void';
    case 'StringLiteralTypeAnnotation':
      return 'string literal';
    case 'NumberLiteralTypeAnnotation':
      return 'number literal';
    case 'BooleanLiteralTypeAnnotation':
      return 'boolean literal';
    case 'NullLiteralTypeAnnotation':
      return 'null';
    case 'NullableTypeAnnotation':
      return `nullable ${getSimplifiedType(removeNullable(annotation))}`;
  }
  throw new Error(annotation.type + ' is not supported');
}

export function compareTypeAnnotation(
  left: AnyTypeAnnotation,
  right: AnyTypeAnnotation,
  isInFunctionReturn: boolean,
): Array<ComparisonResult> {
  switch (left.type) {
    case 'BooleanTypeAnnotation':
      if (
        right.type === left.type ||
        right.type === 'BooleanLiteralTypeAnnotation'
      ) {
        return [];
      }
      return [basicError(left, right, isInFunctionReturn)];
    case 'NumberTypeAnnotation':
      if (
        left.type === right.type ||
        right.type === 'NumberLiteralTypeAnnotation'
      ) {
        return [];
      }
      return [basicError(left, right, isInFunctionReturn)];
    case 'StringTypeAnnotation':
      if (
        left.type === right.type ||
        right.type === 'StringLiteralTypeAnnotation'
      ) {
        return [];
      }
      return [basicError(left, right, isInFunctionReturn)];
    case 'VoidTypeAnnotation':
      if (right.type === 'VoidTypeAnnotation') {
        return [];
      }
      return [basicError(left, right, isInFunctionReturn)];
    case 'StringLiteralTypeAnnotation':
    case 'BooleanLiteralTypeAnnotation':
    case 'NumberLiteralTypeAnnotation':
      if (right.type === left.type) {
        if (right.value === left.value) {
          return [];
        }
        return [literalError(left, right, isInFunctionReturn)];
      }
      return [basicError(left, right, isInFunctionReturn)];
    case 'NullLiteralTypeAnnotation':
      if (right.type !== 'NullLiteralTypeAnnotation') {
        return [basicError(left, right, isInFunctionReturn)];
      }
      return [];
    case 'NullableTypeAnnotation':
      if (right.type === 'NullableTypeAnnotation') {
        return compareTypeAnnotation(
          left.typeAnnotation,
          right.typeAnnotation,
          isInFunctionReturn,
        );
      }
      if (
        right.type === 'NullLiteralTypeAnnotation' ||
        right.type === 'VoidTypeAnnotation'
      ) {
        return [];
      }
      return compareTypeAnnotation(
        left.typeAnnotation,
        right,
        isInFunctionReturn,
      );
    default:
      throw new Error(left.type + 'is not supported');
  }
}

function basicError(
  left: AnyTypeAnnotation,
  right: AnyTypeAnnotation,
  isInFunctionReturn: boolean,
): ComparisonResult {
  const newAnnotationType = isInFunctionReturn ? left : right;
  const oldAnnotationType = isInFunctionReturn ? right : left;
  const newType = getSimplifiedType(newAnnotationType);
  const oldType = getSimplifiedType(oldAnnotationType);
  const reason = isInFunctionReturn
    ? 'is incompatible with what the native code returns'
    : 'native code will break when js calls it';
  return {
    message: `Error: cannot change ${oldType} to ${newType} because ${reason}.`,
    newTypeLoc: newAnnotationType.loc,
    oldTypeLoc: oldAnnotationType.loc,
  };
}

function getValueFromType(annotation: LiteralTypeAnnotation): string {
  if (annotation.type === 'StringLiteralTypeAnnotation') {
    return annotation.value;
  }
  return JSON.stringify(annotation.value);
}

function literalError(
  left: LiteralTypeAnnotation,
  right: LiteralTypeAnnotation,
  isInFunctionReturn: boolean,
): ComparisonResult {
  const newAnnotationType = isInFunctionReturn ? left : right;
  const oldAnnotationType = isInFunctionReturn ? right : left;
  const newType = getSimplifiedType(newAnnotationType);
  const oldType = getSimplifiedType(oldAnnotationType);
  const newValue = getValueFromType(newAnnotationType);
  const oldValue = getValueFromType(oldAnnotationType);
  const reason = isInFunctionReturn
    ? 'is incompatible with what the native code returns'
    : 'native code will break when js calls it';
  return {
    message: `Error: cannot change ${oldType} with value '${oldValue}' to ${newType} with value '${newValue}' because ${reason}.`,
    newTypeLoc: newAnnotationType.loc,
    oldTypeLoc: oldAnnotationType.loc,
  };
}
