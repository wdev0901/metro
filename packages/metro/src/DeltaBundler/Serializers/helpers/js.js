/**
 * Copyright (c) 2015-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 * @format
 */

'use strict';

const addParamsToDefineCall = require('../../../lib/addParamsToDefineCall');
const path = require('path');

import type {Module} from '../../traverseDependencies';

export type Options = {
  +createModuleId: string => number | string,
  +dev: boolean,
};

// Used to include paths in production bundles for traces of performance tuned runs,
// e.g. to update fbandroid/apps/fb4a/compiled_react_native_modules.txt
// Make sure to set PRINT_REQUIRE_PATHS = true too, and restart Metro
const PASS_MODULE_PATHS_TO_DEFINE = false;

function wrapModule(module: Module, options: Options) {
  if (module.output.type === 'script') {
    return module.output.code;
  }

  const moduleId = options.createModuleId(module.path);
  const params = [
    moduleId,
    Array.from(module.dependencies.values()).map(options.createModuleId),
  ];

  // Add the module name as the last parameter (to make it easier to do
  // requires by name when debugging).
  // TODO (t26853986): Switch this to use the relative file path (once we have
  // as single project root).
  if (PASS_MODULE_PATHS_TO_DEFINE || options.dev) {
    params.push(path.basename(module.path));
    if (PASS_MODULE_PATHS_TO_DEFINE) {
      params.push(module.path);
    }
  }

  return addParamsToDefineCall(module.output.code, ...params);
}

module.exports = {
  wrapModule,
};
