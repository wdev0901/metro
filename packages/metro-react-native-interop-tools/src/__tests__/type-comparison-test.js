/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+metro_bundler
 * @flow strict-local
 * @format
 */

import {parse} from 'hermes-parser';
import {getTypeAnnotation} from '../ast-helpers.js';
import {compareTypeAnnotation} from '../type-comparison.js';
import type {AnyTypeAnnotation} from '../type-annotation.js';
import type {FlowType, Statement} from '@babel/types';

function getTypeFromAlias(body: Statement): FlowType {
  if (body.type !== 'TypeAlias') {
    throw new Error('This function support only TypeAlias');
  }
  return body.right;
}

function getTypeFromCode(stringType: string): AnyTypeAnnotation {
  const ast = parse('type T =' + stringType, {
    babel: true,
    sourceType: 'module',
    sourceFilename: 'NativeDeviceManager.js',
  });
  const astType = getTypeFromAlias(ast.program.body[0]);
  return getTypeAnnotation(astType);
}

test.each([
  ['boolean', 'boolean'],
  ['string', 'string'],
  ['number', 'number'],
  ['void', 'void'],
  ['boolean', 'number'],
  ["'a'", "'a'"],
  ["'a'", "'b'"],
  ['8', '8'],
  ['2', '8'],
  ['2', 'number'],
  ['number', '2'],
  ['string', "'a'"],
  ["'a'", 'string'],
  ['true', 'true'],
  ['true', 'false'],
])('comparing basic types', (left, right) => {
  const result = JSON.stringify(
    compareTypeAnnotation(getTypeFromCode(left), getTypeFromCode(right)),
    null,
    '\t',
  );
  expect(`
    left-type:
      ${left}
    right-type:
      ${right}
    output:
      ${result}`).toMatchSnapshot();
});
