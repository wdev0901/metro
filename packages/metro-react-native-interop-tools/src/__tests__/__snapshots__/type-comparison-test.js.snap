// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`comparing basic types 1`] = `
"
    left-type:
      boolean
    right-type:
      boolean
    output:
      no errors"
`;

exports[`comparing basic types 2`] = `
"
    left-type:
      string
    right-type:
      string
    output:
      no errors"
`;

exports[`comparing basic types 3`] = `
"
    left-type:
      number
    right-type:
      number
    output:
      no errors"
`;

exports[`comparing basic types 4`] = `
"
    left-type:
      void
    right-type:
      void
    output:
      no errors"
`;

exports[`comparing basic types 5`] = `
"
    left-type:
      boolean
    right-type:
      number
    output:
      Error: cannot change boolean to number because it will break the native code.
		"
`;

exports[`comparing basic types 6`] = `
"
    left-type:
      'a'
    right-type:
      'a'
    output:
      no errors"
`;

exports[`comparing basic types 7`] = `
"
    left-type:
      'a'
    right-type:
      'b'
    output:
      Error: cannot change string literal to string literal because it will break the native code.
		"
`;

exports[`comparing basic types 8`] = `
"
    left-type:
      8
    right-type:
      8
    output:
      no errors"
`;

exports[`comparing basic types 9`] = `
"
    left-type:
      2
    right-type:
      8
    output:
      Error: cannot change number literal to number literal because it will break the native code.
		"
`;

exports[`comparing basic types 10`] = `
"
    left-type:
      2
    right-type:
      number
    output:
      Error: cannot change number literal to number because it will break the native code.
		"
`;

exports[`comparing basic types 11`] = `
"
    left-type:
      number
    right-type:
      2
    output:
      no errors"
`;

exports[`comparing basic types 12`] = `
"
    left-type:
      string
    right-type:
      'a'
    output:
      no errors"
`;

exports[`comparing basic types 13`] = `
"
    left-type:
      'a'
    right-type:
      string
    output:
      Error: cannot change string literal to string because it will break the native code.
		"
`;

exports[`comparing basic types 14`] = `
"
    left-type:
      true
    right-type:
      true
    output:
      no errors"
`;

exports[`comparing basic types 15`] = `
"
    left-type:
      true
    right-type:
      false
    output:
      Error: cannot change boolean literal to boolean literal because it will break the native code.
		"
`;

exports[`comparing basic types 16`] = `
"
    left-type:
      string
    right-type:
      ?string
    output:
      Error: cannot change string to nullable string because it will break the native code.
		"
`;

exports[`comparing basic types 17`] = `
"
    left-type:
      ?number
    right-type:
      number
    output:
      no errors"
`;

exports[`comparing basic types 18`] = `
"
    left-type:
      ?boolean
    right-type:
      true
    output:
      no errors"
`;

exports[`comparing basic types 19`] = `
"
    left-type:
      ?string
    right-type:
      ?string
    output:
      no errors"
`;

exports[`comparing basic types 20`] = `
"
    left-type:
      ?string
    right-type:
      ?'foo'
    output:
      no errors"
`;

exports[`comparing basic types 21`] = `
"
    left-type:
      ?string
    right-type:
      ?number
    output:
      Error: cannot change string to number because it will break the native code.
		"
`;

exports[`comparing basic types 22`] = `
"
    left-type:
      null
    right-type:
      null
    output:
      no errors"
`;

exports[`comparing basic types 23`] = `
"
    left-type:
      ?string
    right-type:
      null
    output:
      no errors"
`;

exports[`comparing basic types 24`] = `
"
    left-type:
      ?boolean
    right-type:
      void
    output:
      no errors"
`;
